aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -162,10 +171,342 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+  if (blocks_per_slice > 1) {
+
+    if (tidx < RADIX_DIGITS) {
+      counts[block_idx * RADIX_DIGITS + tidx] = digit_count;
+    }
+    __threadfence(); // make sure writes are globally visible
+    __syncthreads(); // make sure all writes are finished before update semaphores
+  }
+
+  // the last block of each slice accumulates counters from multiple blocks and updates desired and ks_to_find
+  __shared__ bool s_is_last_block_done;
+
+  if (tidx == 0) {
+    if (blocks_per_slice == 1) {
+      s_is_last_block_done = true;
+    } else {
+      int blocks_finished_old = atomicAdd(&semaphores[slice_idx], 1);
+      s_is_last_block_done = (blocks_finished_old == blocks_per_slice - 1);
+    }
+  }
+
+  __syncthreads();
+
+  if (!s_is_last_block_done)
+    return;
+
+  // accumulates counters from multiple blocks
+  if (tidx < RADIX_DIGITS && blocks_per_slice > 1) {
+    digit_count = 0;
+    for (int blk = 0; blk < blocks_per_slice; ++blk) {
+      digit_count += counts[(slice_idx * blocks_per_slice + blk) * RADIX_DIGITS + tidx];
+    }
+  }
+
+  // compute the block-wide inclusive prefix sum
+  IndexType& digit_count_cumsum = digit_count;
+  BlockPrefixCallbackOp prefix_op(0);
+  BlockScan(temp_storage.scan_storage).InclusiveSum(digit_count, digit_count_cumsum, prefix_op);
+  __syncthreads();
+  // every thread also need the perfix_sum of it's left value for comparison, so save a copy in shared mem
+  if (tidx < RADIX_DIGITS) {
+    temp_storage.digit_count_cumsum[tidx] = digit_count_cumsum;
+  }
+  __syncthreads();
+
+  if (tidx < RADIX_DIGITS) {
+    IndexType digit_count_cumsum_left = (tidx == 0) ? 0 : temp_storage.digit_count_cumsum[tidx - 1];
+
+    // if not the last pass: update desired and ks_to_find
+    // if last pass: write out the kth value
+    if (digit_count_cumsum_left < k_to_find && k_to_find <= digit_count_cumsum) {
+      desired = at::cuda::Bitfield<Bitwise>::setBitfield(desired, tidx, current_bit, RADIX_BITS);
+      if (current_bit > 0) {
+        desires[slice_idx] = desired;
+        ks_to_find[slice_idx] = k_to_find - digit_count_cumsum_left;
+      } else {
+        kthValues[slice_idx] = TopKTypeConfig<T>::deconvert(desired);
+      }
+    }
+  }
+
+  // reset semaphores for the next pass
+  if (tidx == 0) {
+    semaphores[slice_idx] = 0;
+  }
+};
+
+template <typename T, typename IndexType, int Dim>
+void launch(
+    at::cuda::detail::TensorInfo<T, IndexType> input,
+    IndexType inputSliceSize,
+    IndexType outputSliceSize, // aka `k`
+    bool largest,
+
+    IndexType numInputSlices,
+    IndexType inputWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<T, IndexType> topK,
+    IndexType topKWithinSliceStride,
+
+    at::cuda::detail::TensorInfo<int64_t, IndexType> indices,
+    IndexType indicesWithinSliceStride) {
+
+  // configure items_per_thread based on device architecture and input size
+  // occupancy of this kernel is limited by registers per threads
+  constexpr int REGS_PER_THREAD = 40;
```

Comment:
Could you add a comment describing where this value comes from?

Reply:
added